git/git.go

```diff
@@ -0,0 +1,181 @@
+// It returns the fingerprint of the key the signature was verified
+// with, or an error.
+func (c *Commit) Verify(keyRing ...string) (string, error) {
+	if c.Signature == "" {
+		return "", fmt.Errorf("commit does not have a PGP signature")
+	}
+
+	for _, r := range keyRing {
+		reader := strings.NewReader(r)
+		keyring, err := openpgp.ReadArmoredKeyRing(reader)
+		if err != nil {
+			return "", fmt.Errorf("unable to read armored key ring: %w", err)
+		}
+		signer, err := openpgp.CheckArmoredDetachedSignature(keyring, bytes.NewBuffer(c.Encoded), bytes.NewBufferString(c.Signature), nil)
+		if err == nil && len(signer.PrimaryKey.Fingerprint) >= 20 {
+			return fmt.Sprintf("%X", signer.PrimaryKey.Fingerprint[12:20]), nil
+		}
+	}
+	return "", fmt.Errorf("unable to verify commit with any of the given key rings")
+}
+
+// ShortMessage returns the first 50 characters of a commit subject.
+func (c *Commit) ShortMessage() string {
+	subject := strings.Split(c.Message, "\n")[0]
+	r := []rune(subject)
+	if len(r) > 50 {
+		return fmt.Sprintf("%s...", string(r[0:50]))
+	}
+	return subject
+}
+
+// GitReader knows how to perform read operations for a repository.
+type GitReader interface {
+	// Clone clones a repository from the provided url using the options provided.
+	Clone(ctx context.Context, url string, checkoutOpts CheckoutOptions) (*Commit, error)
+	// IsClean returns whether the working tree is clean.
+	IsClean() (bool, error)
+	// Head returns the hash of the current HEAD of the repo.
+	Head() (string, error)
+	// Path returns the path of the repository.
+	Path() string
+	// Cleanup frees any resources that need to be freed at the end of
+	// the client's lifecycle.
+	Cleanup()
+}
+
+// GitWriter knows how to perform write operations to a repository.
+type GitWriter interface {
+	// Init initializes a repository at the configured path with the origin
+	// remote set to url on the provided branch.
+	Init(ctx context.Context, url, branch string) error
+	// Push pushes the current branch of the repository to origin.
+	Push(ctx context.Context) error
+	// SwitchBranch switches the active branch of the repository to the
+	// provided branch. If the branch doesn't exist, it is created.
+	SwitchBranch(ctx context.Context, branch string) error
+	// WriteFile creates a new file at the given path, reads from the reader and
+	// writes the content to the file. If the file already exists, it's contents
+	// are overwritten.
+	WriteFile(path string, reader io.Reader) error
+	// Commit commits any changes made to the repository.
+	Commit(info Commit, signer *openpgp.Entity) (string, error)
+	// Cleanup frees any resources that need to be freed at the end of
+	// the client's lifecycle.
+	Cleanup()
+}
+
+// GitClient defines all the methods that a git client must cover.
+// An instance of GitClient should have a 1:1 mapping with a Git repository.
+type GitClient interface {
+	GitReader
+	GitWriter
+}
```

Comment:
```suggestion
type Client interface {
	RepositoryReader
	RepositoryWriter
}
```

Reply:
or even `RepositoryClient`, or `Repository`.

Reply:
+1 on `RepositoryReader`/`RepositoryWriter`. As for the `GitClient` interface, I think `RepositoryClient` and `Repository` are better alternatives.

Reply:
Think I am most in favor of `Repository`, as I am not sure the `Client` suffix adds that much value with the reader/writer split.

Reply:
I think the `Client` suffix does add value, because the term `Repository` could hamper readability by misguiding users into thinking this represents a Git repository, where as this is something which can perform operations to a repository.

Reply:
Ack.

Reply:
I may be missing some context of previous discussions but `Client`, `Reader`, `Writer` sound more like how we name things in go, unless there's a possibility of other APIs here that are not about repository. But maybe in that case, these APIs can be a subpackage of git, `git/repository`.